src/main/java/org/candlepin/resource/util/EntitlementEnvironmentFilter.java

```diff
@@ -0,0 +1,147 @@
+ * consumer existing environment. This filtering avoids re-generating all of the
+ * consumer entitlement certificates.
+ */
+public class EntitlementEnvironmentFilter {
+    private EntitlementCurator entitlementCurator;
+    private EnvironmentCurator environmentCurator;
+    private List<String> preExistingEnvs;
+    private List<String> updatedEnvs;
+    private Consumer consumer;
+
+    public EntitlementEnvironmentFilter(EntitlementCurator entitlementCurator,
+        EnvironmentCurator environmentCurator) {
+        this.entitlementCurator = entitlementCurator;
+        this.environmentCurator = environmentCurator;
+    }
+
+    public EntitlementEnvironmentFilter setPreExistingEnvironments(List<String> preExistingEnvs) {
+        this.preExistingEnvs = preExistingEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setUpdatedEnvironment(List<String> updatedEnvs) {
+        this.updatedEnvs = updatedEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setConsumerToBeUpdated(Consumer consumer) {
+        this.consumer = consumer;
+        return this;
+    }
+
+    /**
+     * Returns the list of environment IDs which got added, removed or re-prioritized
+     * from consumers current existing environments.
+     *
+     * @return
+     *  Returns the unique environmentIds which got added, removed or re-prioritized
+     */
+    private List<String> getEnvironmentChangeList() {
+        Set<String> envChangeList = new HashSet<>();
+
+        // Envs Added
+        envChangeList.addAll(CollectionUtils.subtract(this.updatedEnvs, this.preExistingEnvs));
+
+        // Envs removed
+        envChangeList.addAll(CollectionUtils.subtract(this.preExistingEnvs, this.updatedEnvs));
+
+        // Envs reordered (Index within list changes)
+        for (String env : this.updatedEnvs) {
+            if (this.preExistingEnvs.indexOf(env) != this.updatedEnvs.indexOf(env)) {
+                envChangeList.add(env);
+            }
+        }
+
+        return new ArrayList<>(envChangeList);
+    }
+
+    /**
+     * Filters the entitlements who's certificates are needed to be re-generated due to environments
+     * being added, removed or re-prioritized for a consumer.
+     * Idea is to avoid regenerating all of the consumer's entitlements certs & filter the
+     * necessary entitlements who's certificates are actually needed to be re-generated.
+     *
+     * @return
+     *  List of entitlement selected for certificate regeneration
+     */
+    public List<Entitlement> filterEntitlements() {
+        List<Entitlement> entitlements = new ArrayList<>();
+        List<String> environmentChangeList = getEnvironmentChangeList();
+
+        for (Entitlement e : consumer.getEntitlements()) {
+            boolean entToRegen = false;
+            List<String> contentUUIDs = this.entitlementCurator
+                .getProductContentUUIDsByEntitlementId(e.getId());
```

Comment:
Looking at this loop:
- we really don't really need the entitlement objects fetched in memory; we only need their entitlement ids (in order to do further queries)
- Calling `consumer.getEntitlements()` is performing 1 query (because that collection is lazily loaded), and then within the loop we perform N more queries, which is 1 for each entitlement. This is a classic case of N+1, so I would propose to cut this whole thing down to a single query by changing the curator method to something like this:
`Map<String, List<String>> entitlementIdToUUIDList = entitlementCurator.getEntitledContentUUIDsByConsumer(consumer)`
where the maps key is the entitlement id and value is the list of UUIDs for each,
and then iterate through the keys:
```
for (String entitlementID : entitlementIdToUUIDList.keys()) {
    ...
}
```

Reply:
Done.